### PR TITLE
JSONL support

### LINKED ISSUE #448: [CLI Refresh] Output valid JSON and/or JSONL for `workflow list`
Commands such as `workflow list` conceptually output a sequence of objects. Currently `cli-rewrite` JSON output is formatting this as a sequence of multiline JSON objects. This means that the overall result (when there is >1 workflow) is neither valid JSON nor valid [JSONL](https://jsonlines.org/).

A natural requirement is that the CLI should produce either valid JSON or valid JSONL. Possible designs satisfying that requirement:

### A

- `--output json` always produces valid JSON (i.e. we add `[`, commas, and `]` to the `workflow list` output).
- We will also support `--output jsonl` meaning: convert all top-level arrays to newline-separated objects; otherwise do not emit any newlines in JSON.

  Thus `--output jsonl` would apply consistently to all commands (but mainly be useful for commands such as `workflow list` and `env get` that produce top-level arrays, although people could conceivably use it to produce compact output for other commands).

### B

- `--output json` produces valid JSON everywhere except for certain commands (perhaps only `workflow list`) where it produces valid JSONL.

  The only change required here would be eliminating the newlines within `workflow list` items. A downside here is that it will no longer be possible to produce easily human-readable output from `workflow list` without piping into something like `jq`.

### PR DESCRIPTION
**Pull request by cretz**
## What was changed

* Add `--output jsonl` option which removes indentation/newlines from JSON output
* Add printer `StartList` and `EndList`
  * In `--output json` mode, this will add `[` at the beginning, commas before newline of each successive object, and `]` at the end so that it will parse as an array
  * In `--output jsonl` mode, this will do nothing special (indention is already off, just allows normal use of repeated JSON values)
* Added `StartList` and `EndList` to `workflow list`, `batch list`, and `cluster list`

(note, there may be an outstanding flaky batch test we are researching)

## Checklist

1. Closes #448

### CODE REVIEW

File: temporalcli/commands.batch.go
Comment (dandavison) [inline temporalcli/commands.batch.go:75]: I'm thinking maybe we should use the term "array" rather than "list" in the code since that's the correct JSON terminology.

Comment (cretz) [reply temporalcli/commands.batch.go:75]: `list` is what we call the commands and I don't want users to even have to worry about whether JSON is enabled or not. I want people to say whether their jsonl list is starting/ending too. Often people don't think of arrays as lazy/streaming. 

Having said this, if the word "array" is important enough, I can change to get PR approval.

File: temporalcli/internal/printer/printer.go
Comment (Sushisource) [inline temporalcli/internal/printer/printer.go:62]: This (and the if checks below) says to me that rather than using an empty `JSONIndent` as a special indication to do JSONL, we should instead just have the `JSON` field be `JSON | JSONL` and just be explicit about what mode we're in.

Comment (cretz) [reply temporalcli/internal/printer/printer.go:62]: > we should instead just have the `JSON` field be `JSON | JSONL`

More like `JSON | JSONL | Neither`, but it's ugly in Go. I considered this (and I considered two booleans), but `JSONIndent` is how Go does it (where empty string means no multiline, ref https://pkg.go.dev/encoding/json#Encoder.SetIndent). Throughout the Go JSON printing code they check whether indentation is set. And this option gives the caller the freedom to set the indent they want, even though we're the caller. However, if the way Go did this is too confusing, I can make this a Go enum and add a `isJSON()` that checks for those two values.

Comment (Sushisource) [reply temporalcli/internal/printer/printer.go:62]: I'd prefer that I think but it's not a big deal. I'll approve and leave it up to you. Lack up sum types :cry:

**Review by Sushisource — APPROVED** [review]

**Comment by cretz** [comment]
Merging, but I do not hold the use of empty-jsonindent-means-jsonl nor the use of list vs array as a strong opinion if anyone really wants them changed.

File: temporalcli/commands.batch.go
Comment (dandavison) [reply temporalcli/commands.batch.go:75]: Thanks, that's fine, especially if there's a connection with the use of "list" as a verb in the CLI.

**Pull request merged**
